### ISSUE TITLE
Handling of sequences with ambiguous nucleotide symbols

### ISSUE THREAD

**Issue by torognes** [torognes/vsearch#354]
As of version 2.10.3, vsearch handles ambiguous nucleotide symbols as described below.

When aligning, all ambiguous symbols (`RYSWKMBDHVN`) are given a score of 0 no matter which base it is paired with. If two aligned ordinary symbols (`ACGTU`) are identical they are given the score for matches (default +2), while two non-identical ordinary symbols are given the score for mismatches (default -4). The symbols `T` and `U` are considered identical.

When showing an alignment, two identical symbols (ordinary or ambiguous) are shown with a vertical bar (`|`) while other pairs are shown with a space (` `) between them. `T` and `U` are considered identical.

The UIPAC codes for nucleotides, including the ambiguous symbols are shown on the Wikipedia page for [nucleotide acid notation](https://en.wikipedia.org/wiki/Nucleic_acid_notation).

It is relatively simple to change this behaviour as vsearch internally uses a full 16x16 score matrix during alignment and it could easily be changed to handle ambiguous symbols in a more sophisticated manner. The scores have to be integers though.

For example, the scoring of partial matches, e.g. `A` and `W` (representing `A` or `T`), could be given a small positive score. However, matches of ambiguous symbols must not be given a too high score as they could occur rather often by chance.

Also, in the display of an alignment, a partial match, e.g. `A` vs `W` could be shown with a plus sign (`+`) instead of a blank (` `) between the pair of symbols.

The way usearch handles this seems to have changed a bit between versions. There is a brief discussion about it [in the documentation](https://drive5.com/usearch/manual/opt_id.html).

Any suggestions for how this should be changed, or if it should be changed?

**Comment by frederic-mahe** [comment]
Hi, here's my opinion.

We cannot be sure what scoring system `usearch` uses for its ambiguous nucleotide symbols, so it will be difficult to keep in sync with it. However, I think the differences between `usearch` and `vsearch` are in the alignment visualization and identity percentage computation rather than in the scoring system (note that the `usearch` documentation regarding ambiguous nucleotides is in the `-id` section).

>Also, in the display of an alignment, a partial match, e.g. A vs W could be shown with a plus sign (+) instead of a blank () between the pair of symbols.

I agree. A `+` sign could be used when the ambiguous symbols are compatible, and the blank space could be used for mismatches in general.

```sh
vsearch \
    --usearch_global <(printf ">q\nCRA\n") \
    --db <(printf ">t\nCYW\n") \
    --quiet \
    --minseqlength 1 \
    --id 0.3 \
    --alnout -
```
yields:
```
Query >q
 %Id   TLen  Target
 33%      3  t

 Query 3nt >q
Target 3nt >t

Qry 1 + CRA 3
        |  
Tgt 1 + CYW 3

3 cols, 1 ids (33.3%), 0 gaps (0.0%)
```
could yield (R and Y are not compatible, A and W are compatible):
```
Query >q
 %Id   TLen  Target
 66%      3  t

 Query 3nt >q
Target 3nt >t

Qry 1 + CRA 3
        | +  
Tgt 1 + CYW 3

3 cols, 2 ids (66.6%), 0 gaps (0.0%)
```
To summarize, vsearch's scoring model already takes into account ambiguous nucleotides and I don't think it needs to be changed (ambiguous nucleotides are usually rare in NGS sequences or in reference sequences such as 16S). Beyond the visualization of the pairwise alignment, I think it is important to take into account compatible ambiguous nucleotides when computing the **identity** and the **number of mismatches** between two sequences.

**Comment by andzandz11** [comment]
There are ambig. nucleotides in ITS (fung+plant) sequences because there are usually multiple minor sequence variants (incomplete concerted evolution) per individuum (or even major, but then at least post-sequence assembly fails clearly) that are sequenced together. Therefore my need is an algorithm that handles ambig. nucleotides not as a source of error, but rather as having a full biological meaning. That was the reason I stopped susing BLAST and switched to U/VSEARCH.

**Comment by davidealbanese** [comment]
I agree with @frederic-mahe

**Comment by torognes** [comment]
Supported by commit ed376685ecab7eaae2cab584e36861b2631c02e8.

Ambiguous nucleotide symbols (`MRSVWYHKDBN`) will now count as matching to other symbols if they have in common at least one of the nucleotides (`ACGTU`) they represent. For example: `W` will match `A` and `T`, but also any of `MRVHDN`. This will be indicated with a `+` symbol in alignments. Identical matches between any of `ACGTU` will be indicated with a '|' symbol. This is similar to usearch version 8 and later, as well as in line with @frederic-mahe's suggestion. The alignment score for aligning to any ambiguous symbol is still 0.

**Comment by torognes** [comment]
To test it I have used the command

```
vsearch --usearch_global ambiguous.fasta --db ambiguous.fasta --id 0.5 --maxaccepts 0 --maxrejects 0 --strand plus --qmask none --dbmask none --threads 1 --alnout - | less
```

with the input file (`ambiguous.fasta`) below.

```
>A
CTAGCTAGTTGTGCCAAGTAAAAAAAAAAACTAGCTAGTTGTGCCAAGTA
>B
CTAGCTAGTTGTGCCAAGTABBBBBBBBBBCTAGCTAGTTGTGCCAAGTA
>C
CTAGCTAGTTGTGCCAAGTACCCCCCCCCCCTAGCTAGTTGTGCCAAGTA
>D
CTAGCTAGTTGTGCCAAGTADDDDDDDDDDCTAGCTAGTTGTGCCAAGTA
>G
CTAGCTAGTTGTGCCAAGTAGGGGGGGGGGCTAGCTAGTTGTGCCAAGTA
>H
CTAGCTAGTTGTGCCAAGTAHHHHHHHHHHCTAGCTAGTTGTGCCAAGTA
>K
CTAGCTAGTTGTGCCAAGTAKKKKKKKKKKCTAGCTAGTTGTGCCAAGTA
>M
CTAGCTAGTTGTGCCAAGTAMMMMMMMMMMCTAGCTAGTTGTGCCAAGTA
>N
CTAGCTAGTTGTGCCAAGTANNNNNNNNNNCTAGCTAGTTGTGCCAAGTA
>R
CTAGCTAGTTGTGCCAAGTARRRRRRRRRRCTAGCTAGTTGTGCCAAGTA
>S
CTAGCTAGTTGTGCCAAGTASSSSSSSSSSCTAGCTAGTTGTGCCAAGTA
>T
CTAGCTAGTTGTGCCAAGTATTTTTTTTTTCTAGCTAGTTGTGCCAAGTA
>U
CTAGCTAGTTGTGCCAAGTAUUUUUUUUUUCTAGCTAGTTGTGCCAAGTA
>V
CTAGCTAGTTGTGCCAAGTAVVVVVVVVVVCTAGCTAGTTGTGCCAAGTA
>W
CTAGCTAGTTGTGCCAAGTAWWWWWWWWWWCTAGCTAGTTGTGCCAAGTA
>Y
CTAGCTAGTTGTGCCAAGTAYYYYYYYYYYCTAGCTAGTTGTGCCAAGTA
```

**Comment by torognes** [comment]
Fixed in vsearch version 2.13.0, just released.

**Comment by frederic-mahe** [comment]
note: issue covered by our test-suite
